lib/hash_ring/ets.ex

```diff
@@ -287,83 +362,196 @@ defmodule ExHashRing.HashRing.ETS do
         Process.cancel_timer(timer_ref)
-        do_ring_gen_gc(table, ring_gen)
+        :ok = do_ring_gen_gc(state.current_table, ring_gen)
+        :ok = do_ring_gen_gc(state.previous_table, ring_gen)
         ring_gen
       end
 
-    {:reply, {:ok, ring_gens}, %{state | pending_gcs: %{}}}
+    {:reply, {:ok, ring_gens}, %__MODULE__{state | pending_gcs: %{}}}
   end
 
-  def handle_call({:force_gc, ring_gen}, _from, %{pending_gcs: pending_gcs, table: table} = state) do
+  def handle_call({:force_gc, ring_gen}, _from, %__MODULE__{} = state) do
     {reply, pending_gcs} =
-      case Map.pop(pending_gcs, ring_gen) do
+      case Map.pop(state.pending_gcs, ring_gen) do
         {nil, pending_gcs} ->
           {{:error, :not_pending}, pending_gcs}
 
         {timer_ref, pending_gcs} ->
           Process.cancel_timer(timer_ref)
-          {do_ring_gen_gc(table, ring_gen), pending_gcs}
+          :ok = do_ring_gen_gc(state.current_table, ring_gen)
+          :ok = do_ring_gen_gc(state.previous_table, ring_gen)
+          {:ok, pending_gcs}
       end
 
-    {:reply, reply, %{state | pending_gcs: pending_gcs}}
+    {:reply, reply, %__MODULE__{state | pending_gcs: pending_gcs}}
   end
 
-  def handle_info({:gc, ring_gen}, %{pending_gcs: pending_gcs, table: table} = state) do
+  def handle_info({:gc, ring_gen}, %__MODULE__{} = state) do
     pending_gcs =
-      case Map.pop(pending_gcs, ring_gen) do
+      case Map.pop(state.pending_gcs, ring_gen) do
         {nil, pending_gcs} ->
           pending_gcs
 
         {_stale_timer_ref, pending_gcs} ->
-          do_ring_gen_gc(table, ring_gen)
+          :ok = do_ring_gen_gc(state.previous_table, ring_gen)
+          :ok = do_ring_gen_gc(state.current_table, ring_gen)
           pending_gcs
       end
 
-    {:noreply, %{state | pending_gcs: pending_gcs}}
+    {:noreply, %__MODULE__{state | pending_gcs: pending_gcs}}
   end
 
-  defp rebuild(
-         %{nodes: nodes, name: name, table: table, ring_gen: ring_gen, overrides: overrides} =
-           state
-       ) do
-    new_ring_gen = ring_gen + 1
+  ## Private
 
-    :ets.insert(
-      table,
-      for {hash, node} <- Utils.gen_items(nodes) do
-        {{new_ring_gen, hash}, node}
-      end
-    )
+  defp do_find_node(table, gen, num_nodes, key) do
+    hash = Utils.hash(key)
 
-    config =
-      if map_size(overrides) > 0 do
-        overrides =
-          overrides
-          |> Enum.filter(fn {_, values} -> length(values) > 0 end)
-          |> Map.new()
+    case find_next_highest_item(table, gen, num_nodes, hash) do
+      {_, node} ->
+        {:ok, node}
 
-        {table, new_ring_gen, length(nodes), overrides}
-      else
-        {table, new_ring_gen, length(nodes)}
+      _ ->
+        {:error, :invalid_ring}
+    end
+  end
+
+  @spec do_find_nodes_in_table(
+    key :: term(),
+    table :: :ets.tid(),
+    overrides :: override_map(),
+    gen :: generation(),
+    num_nodes :: non_neg_integer(),
```

Comment:
Could we reuse the `table_config` spec here? I don't think we ever actually work on the `num_nodes` directly.
```suggestion
    table_config :: table_config(),
    overrides :: override_map(),
    gen :: generation(),
```